lib/Dialect/TritonNvidiaGPU/Transforms/MMALowering.cpp

```diff
@@ -101,6 +101,70 @@ struct TCGen5MMAScaleSharedToTmemConversion
   }
 };
 
+std::pair<SmallVector<TCGen5CommitOp>, SmallVector<Value>>
+collectCommitOpsAfter(MMAv5OpInterface mmaOp) {
+  SmallVector<TCGen5CommitOp> commitOps;
+  SmallVector<Value> commitPredicates;
+  Operation *nextOp = mmaOp->getNextNode();
+
+  while (nextOp && !isa<MMAv5OpInterface>(nextOp)) {
+    if (auto commit = dyn_cast<TCGen5CommitOp>(nextOp)) {
+      commitOps.push_back(commit);
+      commitPredicates.push_back(commit.getPred());
+    }
+    nextOp = nextOp->getNextNode();
+  }
+
+  return {commitOps, commitPredicates};
+}
+
+void moveDefiningOpsBefore(Value val, Operation *target) {
+  SmallVector<Operation *> toMove;
+  std::function<void(Value)> collectOpsToMove = [&](Value val) {
+    if (auto defOp = val.getDefiningOp()) {
+      if (defOp->getBlock() == target->getBlock() &&
+          target->isBeforeInBlock(defOp)) {
+        for (Value operand : defOp->getOperands()) {
+          collectOpsToMove(operand);
+        }
+        if (llvm::find(toMove, defOp) == toMove.end()) {
+          toMove.push_back(defOp);
+        }
+      }
+    }
+  };
+
+  collectOpsToMove(val);
+
+  for (Operation *op : toMove) {
+    op->moveBefore(target);
+  }
+}
+
+template <typename TCGen5MMAOpTy>
+class MergeCommitIntoMMA : public OpRewritePattern<TCGen5MMAOpTy> {
+public:
+  using OpRewritePattern<TCGen5MMAOpTy>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TCGen5MMAOpTy op,
+                                PatternRewriter &rewriter) const override {
+    auto [commitOps, predicates] = collectCommitOpsAfter(op);
+    if (commitOps.size() == 0) {
+      return llvm::failure();
+    }
+    for (auto [commit, pred] : llvm::zip(commitOps, predicates)) {
+      if (!pred) {
+        pred = rewriter.create<arith::ConstantIntOp>(op.getLoc(), true, 1);
+      }
+      moveDefiningOpsBefore(commit.getBarrier(), op);
+      moveDefiningOpsBefore(pred, op);
+      op.addCompletionBarrier(commit.getBarrier(), pred);
```

Comment:
this pattern is not correct if the mma op predicate is true or equal to commit pred? Because in the lowering the commit is in the `if pred` block

Reply:
For now I'm going to bring up the original logic from https://github.com/triton-lang/triton/pull/7984. It will miss merging a commit predicated with `i == num_k_tiles - 1`, tho.

```
    if (auto commit = dyn_cast<ttng::TCGen5CommitOp>(nextOp)) {
      if ((isConstTrue(mmaPred) && commit.getPred() == nullptr) ||
          equalPred(mmaPred, commit.getPred())) {
        commitOps.push_back(commit);
      }
    }
```

Reply:
yes, basically the pred applied to the commit will be `mma_pred && barrier_pred` so we just need to be sure that `(!barrier_pred) || mma_pred == 1`. 

So in short I agree checking that mma predicate is true is a always safe.

Reply:
fixed